truss/templates/server/model_wrapper.py

```diff
@@ -201,14 +199,28 @@ def try_load(self):
 
         self.set_truss_schema()
 
+    async def try_load(self):
+        await to_thread.run_sync(self._initialize_model)
```

Comment:
will continuing to run in thread keep the same issues we're seeing today with running in thread?

Reply:
hey -- so this just runs "initialize_model", which sets up the path and initializes the model, on a separate thread. IF you do `async def load`, that will not run on a separate loop.

We do that stuff on a separate thread in case the user does some blocking stuff in the __init__ function, we don't want that to cause liveness checks to fail.

Loom: https://www.loom.com/share/4333906b8a9e46768435c61f8901c2dc.

This is not a perfect solution by any means, but is a path to getting users that need this unblocked

Reply:
I'm still a bit concerned with this path blocking the liveness check. How do we document this or what is our recommended usage? Should this be an "undocumented feature" to unblock a special case or a first-class code path?

Reply:
I dug a bit more around. There seems to be really no good way. Even if the liveness endpoint is sync (in which case FastAPI runs it in a thread) the request handling doesn't work because the event loop is blocked. Repro example:
```python
# Filename: test_app.py

import threading
import time
import asyncio
from fastapi import FastAPI
from fastapi.routing import APIRoute
import requests


app = FastAPI()


def status():
    return {"status": "ok"}


async def block_event_loop():
    print("Event loop is now blocked.")
    # Simulate a blocking operation without yielding control
    start_time = time.time()
    while time.time() - start_time < 15:
        # Busy-wait for 15 seconds
        pass
    print("Event loop is now unblocked.")


async def delayed_block():
    # Wait for a short time to allow the server to start
    await asyncio.sleep(1)
    # Start the blocking operation
    await block_event_loop()


@app.on_event("startup")
async def on_startup():
    asyncio.create_task(delayed_block())
    print("Startup Task Started.")


app.router.routes.append(
    APIRoute("/status", status, methods=["GET"])
)

def start_server():
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=8000, log_level="info")


if __name__ == "__main__":
    server_thread = threading.Thread(target=start_server, daemon=True)
    server_thread.start()
    time.sleep(2)

    def make_requests():
        print("Testing synchronous endpoint (/status)...")
        try:
            response = requests.get("http://127.0.0.1:8000/status")
            print("Response from /status:", response.json())
        except Exception as e:
            print("Error calling /status:", e)

    make_requests()

    # Wait for the event loop to unblock
    print("Waiting for event loop to unblock...")
    time.sleep(16)  # Wait more than 15 seconds to ensure the event loop is unblocked

    # Make requests after event loop is unblocked
    print("\nTesting endpoints after event loop is unblocked...")
    make_requests()

```

The only clean way would be a side car process for status/health/live endpoints.

Reply:
I thought about this a little bit more, and think this actually isn't that bad of a problem.

In cartesia's case, they _are_ doing blocking stuff in their `async load` function, but it finishes within the startup time checks, so the pods don't get killed. In fairness, the pod is _not_ live yet.

Going forward, we are going to have custom health checks, and alongside the ability for users to adjust their health check windows, so they can adjust things accordingly.